Raise Transitive Project References to MSBuild elements

natidea opened the pull request:
Raise closure of Transitive Project References to msbuild as `ProjectReference` elements. This closure is generated by NuGet in the assets file.

Fixes #200 
See https://github.com/dotnet/roslyn-project-system/issues/199#issuecomment-261368911 for details on P2P plan.

natidea committed:
Add targets to include transitive project references before ResolveProjectReferences is called

natidea committed:
Add assets and scaffolding for test to build App With Transitive ProjectRefs

natidea committed:
Add condition to IncludeTransitiveProjectReferences to skip during projects with ProducesNoOutput settings. (Fixes failures building xUnit.net.csproj)

natidea committed:
Fixup Transitive Project Ref tests

natidea committed:
Add Clean target test for AppWithTransitiveProjectRefs

natidea commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.PackageDependencyResolution.targets:
```diff
@@ -261,6 +261,37 @@ Copyright (c) .NET Foundation. All rights reserved.
+  <Target Name="IncludeTransitiveProjectReferences"
+          Condition="'$(ResolvePackageDependenciesForBuild)' == 'true' and '$(EnableResolvePackageDependencies)' == 'true' and Exists('$(ProjectAssetsFile)')"
+          DependsOnTargets="_ComputeTransitiveProjectReferences" 
+          BeforeTargets="AssignProjectConfiguration">
```
`BeforeTargets="AssignProjectConfiguration"` ensures `ProjectReference` elements are created before they are processed into ResolvedProjectReferences. 

Alternative was to add this to the list of `ResolvePackageDependenciesForBuildDependsOn`, but this calls the target too late. However this means we have to copy the condition from `ResolvePackageDependenciesForBuild` to ensure we are not calling `RunResolvePackageDependencies` at inappropriate times. (E.g. Without this condition, the sdk build fails on building [xUnit.net.csproj](https://github.com/dotnet/sdk/blob/master/src/Dependencies/xUnit.net/xUnit.net.csproj)

Other alternative is to bypass ResolveProjectReferences and generate the `_ResolvedProjectReferencePaths` element directly, as is done in [Microsoft.NuGet.targets](https://github.com/NuGet/NuGet.BuildTasks/blob/dev/src/Microsoft.NuGet.Build.Tasks/Microsoft.NuGet.targets#L193). This currently does not work because of a bug calculated resolved paths (TODO insert bug).

dsplaisted replied:
Could we add `IncludeTransitiveProjectReferences` to `ResolvePackageDependenciesForBuildDependsOn`, and then put `BeforeTargets="AssignProjectConfiguration"` on the `ResolvePackageDependenciesForBuild` target?  It seems like that would prevent us from having to repeat the same condition twice.

In any case, I would like to see comments in the targets file itself explaining how the ordering works.  IIUC, the key is that the `AssignProjectConfiguration` is the entry point that reads `ProjectReference` items, so this target needs to run before that, but only if there is an assets file.

Separately, what is the difference between the `ResolvePackageDependenciesForBuild` and `EnableResolvePackageDependencies` properties?  They seem to only be used together.

natidea replied:
It appears the two flag properties are redundant, and may have been added separately. I've removed `EnableResolvePackageDependencies` and also simplified the ordering as you suggested to avoid repeating the condition.

natidea commented:
/cc @dotnet/project-system 
@emgarten for NuGet fyi.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.PackageDependencyResolution.targets:
```diff
@@ -261,6 +261,37 @@ Copyright (c) .NET Foundation. All rights reserved.
+          DependsOnTargets="RunResolvePackageDependencies"
+          Returns="_TransitiveProjectReferences">
+    <ItemGroup>
+      <ProjectDefs Include="@(PackageDefinitions->WithMetadataValue('Type', 'project'))" />
```
Nit: Let's be consistent on not abbreviating things. We have PackageDefinitions and ProjectDefs on the same line here. It should be ProjectDefinitions.

Item names are also supposed to be singular, but that's one Microsoft has gotten wrong in many places so I feel less strongly about enforcing it.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.PackageDependencyResolution.targets:
```diff
@@ -261,6 +261,37 @@ Copyright (c) .NET Foundation. All rights reserved.
+          BeforeTargets="AssignProjectConfiguration">
+    <ItemGroup>
+      <!-- Add the references we computed -->
+      <ProjectReference Include="@(_TransitiveProjectReferences)" />
```
Is it possible for there to be a transitive reference from the assets file to a project that is also directly referenced?  It would be good to have a test case to cover this, where B references C, and A references both B and C.

natidea committed:
Remove irrelevant test scenarios

natidea committed:
Simplify ordering of targets associated with resolving package dependencies and remove redundant condition EnableResolvePackageDependencies

natidea committed:
Merge remote-tracking branch 'dotnet/master' into transitiveProjectRef

natidea committed:
Exclude top level project references from _TransitiveProjectReferences

natidea committed:
Compute transitive project refs from target tree.
Use package dependencies (under "target") instead of package definitions (under "library") to compute transitive project references. Package dependencies could contain different information for different TFMs. This also simplifies determining which project references are transitive since we can do that all using information in the package dependency tree.

natidea commented:
After chatting with @emgarten , I've made some changes. Previously we were computing transitive project references from `ProjectDefinitions` which comes from the asset file's "libraries" section. This is a deduped list of packages/projects. This is not correct because project transitivity could be conditional on the TFM. In this case, the project reference may be handled transitively in one TFM, but not the other.

Commit https://github.com/dotnet/sdk/pull/478/commits/f3b057b3f43e89cf770abf705dfb896faef8ef1c changes to using PackageDependencies so computing transitive project references matches the information for the active target. This includes excluding projects marked with a placeholder compile time assembly `_._`

Additionally I added logic to exclude directly referenced projects to address the concern @dsplaisted raised. It turns out these duplicated projects were deduped by `ResolveAssemblyReference` because their output assemblies were identical, so the compiler only gets one copy. However the Build Hierarchy printed at the end of a diagnostic build contains more entries without this fix, so perhaps more redundant NoOp builds were happening.

I've tested this using the following project layout:
```
    A -> B -> C -> D
              C -> E
    A -> D
```
When building A:
- C and E are treated as transitive references 
- B and D are direct references and hence are already included; they are not appended by the new target

I simulated making E a non-transitive reference in C, by modifying the assets file to insert a placeholder in E. With this change only C is treated as a transitive reference and E is not sent to the compiler. This compiles fine, but I get an error in the [`GenerateBuildDependencyFile`](https://github.com/dotnet/sdk/blob/818f8fdcd5a3dc93fefac8576a584efb258bd5a6/src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets#L71) target about missing project information. We will need to fix [DependencyContextBuilder](https://github.com/dotnet/sdk/blob/aca5d1556f7c9b3e2182a38844d66be663ff26fb/src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs#L424) to handle this case. I'll file a bug on this.

natidea commented:
/cc @srivatsn for Ask Mode approval

natidea committed:
Add unit test for transitive project reference changes in ResolvePackageDependencies task

natidea commented:
Merging so we can get some insertions going. Additional feedback still welcome.

AlexeyRipenko commented:
When this fix will be available in VS 2017?

dsplaisted commented:
It should already be in the latest RC of VS 2017.

KshitizGIT commented:
Any way I can opt out of this?

Lets say, I have the following project layout.

A -> B -> C.

I don't want classes inside assembly C to be directly accessible inside assembly A.

srivatsn commented:
Adding PrivateAssets=All to the ProjectReference stops the flow of the transitive ref. So, in this case. B.csproj would have this:

```xml
  <PackageReference Include="C.csproj" PrivateAssets="All" />
```

nguerrera commented:
@srivatsn, you mean ProjectReference, right?

srivatsn commented:
Ah yes typo - fixed it the original comment.

KshitizGIT commented:
Thanks a lot.. Worked like a charm :+1:
